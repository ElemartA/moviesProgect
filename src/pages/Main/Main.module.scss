
.container {
  max-width: 1500px;
  margin: 0 auto;
  .container__sort {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }
  .wrap {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
  }
}



@media (max-width: 625px) {
  .container {
    max-width: 570px;
    .container__sort {
      margin: 0 auto;
      width: 500px;
  }
  }
}
@media (max-width: 500px) {
  .container {
    width: 410px;
  .container__sort {
      margin: 0 auto;
      width: 400px;
  }
}
  
}
@media (max-width: 320px) {
  .container {
    width: 315px;
  .container__sort {
      margin: 0 auto;
      width: 315px;
  }
}
  
}