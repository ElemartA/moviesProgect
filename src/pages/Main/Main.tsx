import React, { useEffect } from "react";
import OneCard from "../../components/Card/OneCard";
import Search from "../../components/Search";
import s from "./Main.module.scss";
import Loading from "../../components/Loading";
import PageNumber from "../../components/PageNumber";
import Sort from "../../components/Sort/Sort";
import TypeFilm from "../../components/TypeFilm/TypeFilm";
import YearFromTo from "../../components/Years/YearFromTo";
import { useNavigate } from "react-router-dom";
import { defaultState, mainSlice } from "../../store/reducers/MainSlice";
import { useAppDispatch, useAppSelector } from "../../hooks/redux";
import { fetchMain } from "../../store/reducers/ActionCreators";

const Main = () => {
  const {
    filterMovies,
    numberPage,
    sort,
    typeFilm,
    yearFrom,
    yearTo,
    value,
    totalPages,
    isLoading,
    error,
  } = useAppSelector((state) => state.mainReducer);
  const {
    update_number_page,
    update_sort,
    update_type_film,
    update_year_from,
    update_year_to,
    update_value,
    update_id_film,
  } = mainSlice.actions;
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const getMovies = async () => {
    dispatch(
      fetchMain({
        typeFilm: typeFilm,
        yearFrom: yearFrom,
        yearTo: yearTo,
        sort: sort,
        value: value,
        numberPage: numberPage,
      })
    );
  };

  useEffect(() => {
    getMovies();
  }, [numberPage, sort, typeFilm, yearFrom, yearTo]);

  const handleSubmit = async (event: { preventDefault: () => void }) => {
    event.preventDefault();
    getMovies();
    dispatch(update_number_page(defaultState.numberPage));
  };

  const newValue = async (text: string) => {
    dispatch(update_value(text));
  };

  const newPage = async (num: string) => {
    dispatch(update_number_page(num));
  };

  const newSort = (value: string) => {
    dispatch(update_sort(value));
  };

  const newTypeFilm = (value: string) => {
    dispatch(update_type_film(value));
  };

  const newYearFrom = (value: number) => {
    dispatch(update_year_from(value));
  };

  const newYearTo = (value: number) => {
    dispatch(update_year_to(value));
  };

  return (
    <div data-testid="mainPage" className={s.container}>
      <Search
        value={value}
        onChange={(value: string) => {
          newValue(value);
        }}
        onSubmit={handleSubmit}
      />
      <div className={s.container__sort}>
        <Sort
          newSort={(value: string) => {
            newSort(value);
          }}
        />
        <TypeFilm
          newTypeFilm={(value: string) => {
            newTypeFilm(value);
          }}
        />
        <YearFromTo
          newYearFrom={(value: number) => newYearFrom(value)}
          newYearTo={(value: number) => newYearTo(value)}
        />
      </div>
      {isLoading && <Loading />}
      {error.length === 0 ? (
        <div className={s.wrap}>
          {filterMovies.map((item, index) => (
            <OneCard
              key={index}
              nameRu={item.nameRu}
              year={item.year}
              ratingKinopoisk={item.ratingKinopoisk}
              img={item.posterUrlPreview}
              onClick={() => {
                dispatch(update_id_film(item.kinopoiskId));
                navigate("/film" + "/" + item.kinopoiskId);
              }}
            />
          ))}
        </div>
      ) : (
        <p
          style={{
            marginLeft: "50%",
            marginTop: "15px",
            color: "black",
            fontSize: 20,
          }}
        >
          {error}
        </p>
      )}
      <PageNumber
        onClick={(numb: string) => {
          newPage(numb);
        }}
        value={numberPage}
        pageCount={totalPages}
      />
    </div>
  );
};

export default Main;
