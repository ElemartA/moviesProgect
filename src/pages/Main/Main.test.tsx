import React from "react";
import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Main from "./Main";
import axios from "axios";

Object.defineProperty(window, "MutationObserver", {
  writable: false,
  configurable: false,
  value: MutationObserver,
});

jest.mock("axios");

describe("Main", () => {
  it("renders Main component", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Main />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    screen.debug();
    expect(screen.getByRole("img")).toBeInTheDocument();
    expect(screen.queryByText(/Главная/i)).toBeNull();
  });
});

// const getData = async () => {
//   try {
//     const response = await axios.get(
//       `https://kinopoiskapiunofficial.tech/api/v2.2/films/?type=ALL&order=RATING&keyword=человек&page=1`,
//       {
//         headers: {
//           "X-API-KEY": "52b4d23c-d7de-44dc-8d97-4d0bbe579e7c",
//           "Content-Type": "application/json",
//         },
//       }
//     );
//     const years: number[] = response.data.items.map(
//       (item: { year: number }) => item.year
//     );
//     return years;
//   } catch (err) {
//     console.error(err);
//   }
// };

describe("api test", () => {
  let response: {
    data: {
      items: [
        {
          year: string;
          nameRu: string;
        },
        {
          year: string;
          nameRu: string;
        }
      ];
    };
  };
  beforeEach(() => {
    response = {
      data: {
        items: [
          {
            year: "2020",
            nameRu: "Жизнь человека. Последнее интервью",
          },
          {
            year: "2015",
            nameRu: "Человек",
          },
        ],
      },
    };
  });
  it("тест компонента с асинхронной загрузкой данных с сервера", async () => {
    (axios.get as jest.Mock).mockReturnValue(response);
    render(<Main />);

    const authors = await screen.findAllByTestId("year-id");
    expect(authors.length).toBe(2);
    expect(axios.get).toBeCalledTimes(2);
    screen.debug();
  });
});
