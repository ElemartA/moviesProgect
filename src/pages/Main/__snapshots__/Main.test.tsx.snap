// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders Main component 1`] = `
<DocumentFragment>
  <div>
    <div
      class="preloader"
    >
      <img
        alt="preloader"
        src="/preloader.svg"
        width="50"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`api test тест компонента с асинхронной загрузкой данных с сервера 1`] = `
<DocumentFragment>
  <div>
    <div
      class="preloader"
    >
      <img
        alt="preloader"
        src="/preloader.svg"
        width="50"
      />
    </div>
  </div>
</DocumentFragment>
`;
